marlowe-playground-client/src/Wallet.purs

```diff
@@ -0,0 +1,1301 @@
+render state =
+  div []
+    [ section [ classes [ panelSubHeader, aHorizontal ] ]
+        [ div [ classes [ panelSubHeaderMain, aHorizontal ] ]
+            [ div [ classes [ ClassName "wallet-title", aHorizontal, jFlexStart ] ]
+                [ div [ classes [ ClassName "demos", spaceLeft ] ]
+                    [ small_ [ text "Created Wallets:" ]
+                    ]
+                ]
+            , ul [ classes [ ClassName "wallet-list", aHorizontal ] ]
+                (map (displayWallet walletName) wallets)
+            , addWalletButton
+            , a [ class_ (ClassName "clear-all-contracts"), onClick $ Just <<< const ResetAll ] [ text "Clear All" ]
+            ]
+        , div [ classes [ panelSubHeaderSide, expanded (state ^. _showRightPanel) ] ]
+            [ a [ classes [ (ClassName "drawer-icon-click") ], onClick $ const $ Just $ ShowRightPanel (state ^. (_showRightPanel <<< to not)) ]
+                [ img [ src closeDrawerIcon, class_ (ClassName "drawer-icon") ] ]
+            ]
+        ]
+    , section [ class_ (ClassName "code-panel") ]
+        [ mainPanel state
+        , rightPanel state
+        ]
+    ]
+  where
+  isRunning = fromMaybe false $ state ^? (_walletLoadedContract <<< _Just <<< _started)
+
+  walletName = fromMaybe "No Wallet Open" $ state ^? (_openWallet <<< _Just <<< _name)
+
+  wallets = map snd $ Map.toUnfoldable $ state ^. _wallets
+
+  addWalletButton = button [ class_ (ClassName "add-wallet-button"), onClick $ Just <<< const CreateWallet ] [ text "+" ]
+
+displayWallet :: forall p. PubKey -> Wallet -> HTML p Action
+displayWallet openWallet wallet =
+  li
+    [ class_
+        ( if wallet ^. _name == openWallet then
+            active
+          else
+            ClassName mempty
+        )
+    ]
+    [ a [ onClick $ const $ Just $ SelectWallet wallet ] [ text $ show wallet ]
+    ]
+
+mainPanel ::
+  forall p.
+  State ->
+  HTML p Action
+mainPanel state =
+  div [ class_ (ClassName "wallet-contents") ] case state ^. _view of
+    Home -> []
+    WalletView walletName ->
+      ( [ div [ class_ (ClassName "wallet-top-section") ]
+            [ renderAssets state
+            , renderMyRoles walletName state
+            ]
+        , hr_
+        , renderContracts state
+        , hr_
+        ]
+          <> renderContract'
+      )
+  where
+  isStarted = fromMaybe false $ state ^? (_walletLoadedContract <<< _Just <<< _started)
+
+  isLoaded = state ^. (_walletLoadedContract <<< to isJust)
+
+  renderContract'
+    | not isLoaded = []
+
+  renderContract'
+    | not isStarted =
+      [ button
+          [ classes [ ClassName "tooltip", ClassName "start-simulation-btn" ]
+          , onClick $ const $ Just $ StartContract
+          ]
+          [ span [ class_ (ClassName "tooltiptext") ] [ text "Start Contract" ]
+          , text "Start"
+          ]
+      , div [ classes [ ClassName "code-editor", ClassName "expanded" ] ]
+          [ pre [] [ text contractString ] ]
```

Comment:
`pre_`?

Reply:
I always start out with the 2 arg versions and then reduce (or forget to). Maybe I should consider doing the opposite 🤔